File: apps/api/app/schemas/cockpit_mode.py
"""Schemas for the cockpit mode selector API."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

_MODE_IDS = {
    "learning",
    "manual",
    "auto_paper",
    "assisted_live",
    "live",
    "auto_live",
}


class CockpitModeOptionSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    label: str
    status: Literal["active", "available", "locked"]
    selectable: bool
    locked: bool
    reason: str
    risk_note: str
    allowed_actions: list[str] = Field(default_factory=list)
    blocked_actions: list[str] = Field(default_factory=list)
    safety_gates: list[str] = Field(default_factory=list)

    @field_validator("id")
    @classmethod
    def validate_id(cls, value: str) -> str:
        normalized = value.strip().lower()
        if normalized not in _MODE_IDS:
            raise ValueError(f"Unsupported cockpit mode: {value}")
        return normalized


class CockpitModeSafetyStateSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    live_trading_enabled: bool
    auto_live_enabled: bool
    real_money_enabled: bool
    paper_order_submission_allowed: bool
    live_order_submission_allowed: bool
    auto_trading_allowed: bool
    emergency_stop_active: bool
    trading_mode: str
    execution_control: str
    arming_state: str
    reasons: list[str] = Field(default_factory=list)


class CockpitModeResponseSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    current_mode: str
    selectable_modes: list[str] = Field(default_factory=list)
    locked_modes: list[str] = Field(default_factory=list)
    modes: list[CockpitModeOptionSchema] = Field(default_factory=list)
    global_safety_state: CockpitModeSafetyStateSchema
    live_trading_enabled: bool
    auto_live_enabled: bool
    real_money_enabled: bool
    notes: list[str] = Field(default_factory=list)

    @field_validator("current_mode")
    @classmethod
    def validate_current_mode(cls, value: str) -> str:
        normalized = value.strip().lower()
        if normalized not in _MODE_IDS:
            raise ValueError(f"Unsupported cockpit mode: {value}")
        return normalized


class CockpitModeUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    requested_mode: str

    @field_validator("requested_mode")
    @classmethod
    def validate_requested_mode(cls, value: str) -> str:
        normalized = value.strip().lower()
        if normalized not in _MODE_IDS:
            raise ValueError(f"Unsupported cockpit mode: {value}")
        return normalized

File: apps/web/lib/api/cockpitMode.ts
import { apiRequest } from "./core";

export type CockpitModeId =
  | "learning"
  | "manual"
  | "auto_paper"
  | "assisted_live"
  | "live"
  | "auto_live";

export interface CockpitModeOption {
  id: CockpitModeId;
  label: string;
  status: "active" | "available" | "locked";
  selectable: boolean;
  locked: boolean;
  reason: string;
  risk_note: string;
  allowed_actions: string[];
  blocked_actions: string[];
  safety_gates: string[];
}

export interface CockpitModeSafetyState {
  live_trading_enabled: boolean;
  auto_live_enabled: boolean;
  real_money_enabled: boolean;
  paper_order_submission_allowed: boolean;
  live_order_submission_allowed: boolean;
  auto_trading_allowed: boolean;
  emergency_stop_active: boolean;
  trading_mode: string;
  execution_control: string;
  arming_state: string;
  reasons: string[];
}

export interface CockpitModeResponse {
  current_mode: CockpitModeId;
  selectable_modes: CockpitModeId[];
  locked_modes: CockpitModeId[];
  modes: CockpitModeOption[];
  global_safety_state: CockpitModeSafetyState;
  live_trading_enabled: boolean;
  auto_live_enabled: boolean;
  real_money_enabled: boolean;
  notes: string[];
}

interface CockpitModeErrorDetail {
  code?: string;
  message?: string;
}

function parseCockpitModeError(message: string): string {
  const marker = message.indexOf(": {");
  if (marker === -1) {
    return message;
  }

  try {
    const parsed = JSON.parse(message.slice(marker + 2).trim()) as {
      detail?: CockpitModeErrorDetail;
    };
    return parsed.detail?.message ?? message;
  } catch {
    return message;
  }
}

export async function getCockpitMode(): Promise<CockpitModeResponse> {
  return apiRequest<CockpitModeResponse>("/cockpit/mode", { method: "GET" });
}

export async function updateCockpitMode(
  requested_mode: CockpitModeId,
): Promise<CockpitModeResponse> {
  try {
    return await apiRequest<CockpitModeResponse>("/cockpit/mode", {
      method: "POST",
      body: JSON.stringify({ requested_mode }),
    });
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(parseCockpitModeError(error.message));
    }
    throw error;
  }
}